### PR TITLE
"Document your tag API" rule

### PR DESCRIPTION
**Pull request by jbmoelker**
## Summary

Proposed changes:
- "Document your tag API" rule (incl. summary, why, how and example).
- This section should be just before the "Add a tag demo" rule.
## Checklist
- [X] The changes only affect or contain one style guide rule or section.
- [X] The changes are in [style guide format](/CONTRIBUTING.md#format).
- [X] The new style guide section has an entry in the [Table of Contents](/README.md#table-of-contents) (only if changes introduce new section).
- [X] The changes can be merged into the target branch without conflicts.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -224,3 +225,50 @@ my-example li { }
...
+
+
+## Document your tag API
+
+A Riot tag instance is created by using the tag element inside your application. The instance is configured through its custom attributes. For the tag to be used by other developers, these custom attributes - the tag's API - should be documented in a `README.md` file.
+
+### Why?
+
+* Documentation provides developers with a high level overview to a module, without the need to go through all its code. This makes a module more accessible and easier to use.
+* A tag's API is the set of custom attributes through which its configured. Therefore these are especially of interest to other developers which only want to consume (and not develop) the tag.
```
Comment (thadeetrompetter) [inline README.md]: it's

File: README.md [hunk 2]
```diff
@@ -224,3 +225,50 @@ my-example li { }
...
+Within the README file, describe the functionality and the usage of the module. For a tag module its most useful to describe the custom attributes it supports as those are its API:
+
+```markdown
+# Range slider
+
+## Functionality
+
+The range slider lets the user to set a numeric range by dragging a handle on a slider rail for both the start and end value.
+
+This module uses the [noUiSlider](http://refreshless.com/nouislider/) from cross browser and touch support.
```
Comment (thadeetrompetter) [inline README.md]: from = for?

**Comment by phortuin** [comment]
I don't agree with the README example. I have noticed that when components grow larger, the documentation should become better-structured. I have examples I can share that we maybe should  make into a separate README style guide, and point to there from here.

**Comment by phortuin** [comment]
# Single select

The Single select is a dropdown menu which allows the user to select a single item from a stylized dropdown list. If the HTML element the SingleSelect is invoked in is a The component's options will be constructed from the options from the select. If SingleSelect is invoked on an arbitrary element, you'll have to provide an array of objects, which will become the dropdown's menu items.

> ### Important Guidelines
  - Should be used inside a `<form>` element
  - If possible, should be implemented as an actual HTML `<select>` with `<option>`s which still works without JavaScript. You can find examples throughout the documentation.
  - When using this component as an autocomplete make sure the global variable `pathToAssets` in the `<head>` is set to point to the assets folder (defaults to `../../assets/`).

<div class="guide-example">
	<div class="guide-panel">
		<span class="extended-select single-select">
			<span href="#" role="button" class="extended-select-button">
				<span class="extended-select-status has-placeholder">Select an option</span>
				<span class="extended-select-icon"></span>
			</span>
		</span>
	</div>
</div>

``` markup
<span class="extended-select single-select">
	<span href="#" role="button" class="extended-select-button">
		<span class="extended-select-status has-placeholder">Select an option</span>
		<span class="extended-select-icon"></span>
	</span>
</span>
```

**Note** that the example above is only a flat, non-working representation of a select box due to limitations in the documentation code. See the demo page for a live use case.

## Usage

Initialize like this:

``` javascript
SingleSelect(
	'[data-extended-select="example-select"]',
	{ /* options */ },
	[ /* data */ ]
);
```

| Argument&nbsp;nr. | Type | Description |
| --------------- | ---- | ----------- |
| 1 | string &verbar; jQuery object | The selector, or a jQuery object, of the element that should become a single select |
| 2 | object | Hash of options, explained below |
| 3 | array | Array with data, usually consists of JavaScript object literals; ex: `[ { id: 'nl', label: 'Netherlands' }, { id: 'be', label: 'Belgium' } ]` |

## Configuration

There are two ways of configuring the single select: by providing an options hash when initializing in JavaScript, or by using `data` attributes on the element that you are initializing the single select on:

``` markup
<select data-width="220px" 
		data-template="<span class='country'>{{name}}</span>" 
		data-value-reference="isoCode">
</select>
```

For options with boolean values, attaching just the attribute name (without a value) on the element will be equal to explicitly setting the attribute value to `true`:

``` markup
<select data-autocomplete
		data-can-add-item>
</select>
```

**Note** that if you use pipe-separated values in a `data`-attribute, the result will be an array:

``` markup
<select data-search-values="isoCode|country"></select>
```

&hellip;will be interpreted as an array in JavaScript: `[ 'isoCode', 'country' ]`.

## Options

All the parameters below are camelCase; to use the equivalent as a `data`-attribute, translate to dash-case (ex: `isoCode => iso-code`).

| Parameter         | Type    | Default     | Description                      |
| ----------------- | ------- | ----------- | -------------------------------- |
| autocomplete | boolean | false | Whether or not the single select will behave like an autocomplete select box. See below for more information | 
| dropDownWidth| string | '100%' | Width (in any valid CSS size unit) of the drop down box. Any width set in percentages relates to the parent element, not the width of the input/select element itself. **Note** that when left unchanged, it might adapt to the setting of `width`. |
| keyReference |  string | 'id' | Defines what property in the dataset is the key or id of the item. For example, an item with the following JSON: `{ "isoCode": "NL", "country": "Netherlands" }` can have the keyReference set to 'isoCode' |
| placeholder | string | _depends on `autocomplete` setting_ | The placeholder text that will be shown on the element when nothing is selected. If not specified and autocomplete is true it will default to 'Type to find results', else 'Select an option' |
| searchValues | array | [] | An array of properties which are used (in order) to match what is typed in the dropdown to jump to the corresponding item. The properties exist in the dataset ofcourse. Will default to the `valueReference`property. See below for an example and more explanation. |
| template | string | '&lt;span>{{&nbsp;label&nbsp;}}&lt;/span>' | The template to display each of the dropdown items in. See below for more details |
| valueReference | string | 'label' | Defines what property in the dataset is the value of the item. Works in tandem with the option `keyReference` |
| width | string &verbar; false | false | Width (in any valid CSS size unit) of the select element. If false, its width will not be specified; this resolves to `auto` or what is defined in CSS. **Note** that if you set this option, and `dropDownWidth` is not set, the width of the drop down will match the width set here. |

## Options — only for autocomplete

| Parameter         | Type    | Default     | Description                      |
| ----------------- | ------- | ----------- | -------------------------------- |
| canAddItem | boolean | false | If true, an extra option appears in the autocomplete-dropdown that enables users to select whatever they typed as a new value that will be stored in the database |
| clearOnFocus | boolean | false | If true, any typed text in the autocomplete search field will be cleared when it gets focus |
| delay | number | 500 | The delay in milliseconds after the last keypress before firing the autocomplete event |
| maxInputLength | number | 35 | The maximum number of characters that can be typed in an autocomplete search input |
| maxSearchResults | number &verbar; false | false | The maximum number of search results in an autocomplete-dropdown. Either false (no limit) or a specific number | 
| moreResultsText | string | 'More results found' | Text that will be displayed when more results exist, but can not be displayed due to `maxSearchResults` setting |
| noResultsText | string | 'No results found' | Text that will be displayed when there are no results for the query |
| triggerInputLength | number | 3 | Minimum number of characters that need to be typed before `onAutocompleteQuery` event triggers |

### SearchValues

If a user types in a (non-autocomplete) dropdown while it has focus, the selection will jump to the first matching item. For example, if item 1 is 'Belgium' and item 2 is 'Netherlands', typing the letter 'N' will jump to item 2. 

The default behaviour of this 'jump' is using the `valueReference` property to know what string to match.

Setting the `searchValues` option can change this behaviour. You can pass an array with properties that will form a string to match. For example, passing `[ 'isoCode', 'language' ]` as the searchValues option will result in strings like `NL Dutch` and `ES Spanish`. In this case, typing `es` will match the item representing Spain.

**Note** that matching the user input to the `searchValue` is not dependant on what template you use / what the user sees in the drop down.

### Autocomplete
Autocomplete allows the user to search for a pre-existing value. Usually, this data comes from an API call. The interaction model for this is based on the idea that there are too many options for a normal select field to be usable, and therefore making a selection is done after filtering the results by entering part of the desired option's text. Example:

- User types `ams`
- Database returns 2 options: `Amstelveen`, `Amsterdam` and shows them in the dropdown 
- User selects `Amsterdam`
- Select box represents this selection by showing `Amsterdam` in the text input area.

### CanAddItem
If it's allowed for a user to add an item that is not yet in the database, the option `canAddItem` should be passed when creating the Single Select. The following scenario is now possible:

- User types `Ams`
- Database returns 2 options: `Amstelveen`, `Amsterdam` and shows them in the dropdown, along with `Use Ams`
- User types `Amstelhoek`
- Dropdown shows only `Use Amstelhoek`
- User selects `Use Amstelhoek`
- Select box shows `Amstelhoek` in text input area

### Template
The template is simply HTML with a Mustache-like syntax. Variables that exist in the dataset are injected with double curly brackets: `{{ variable }}`. You should wrap in at least a `<span>`. Here is an example of a proper template:

``` markup
<span><strong>{{ country }}</strong>( {{ isoCode }} )</span>
```

## Adding data

There are several ways to add data (meaning, options) to the single select. 

- By using `<option>` elements in HTML. This is the most straightforward and semantically correct option;
- By providing an array with data directly when initializing the component;
- By doing an asynchronous API call and setting data on the component via the `setData` function;
- By doing an asynchronous API call and setting data on the component via the `setAutocompleteResult` function; this will only work for single selects with autocomplete enabled.

For the first three methods, here are minimal implementations that will yield the same result:

### Adding data via &lt;option>

``` markup
<select id="my-select">
	<option value="nl">Netherlands</option>
	<option value="be">Belgium</option>
</select>
```

``` javascript
SingleSelect('#my-select');
```

**Notes:**

- The text of the `<option>` is stored internally under the property `label`.
- The first `<option>` in the example above is selected automatically.
- Since HTML knows only strings, your return object when a user selects something will only have strings, not numbers.

### Adding data when initializing

``` markup
<select id="my-select"></select>
```

``` javascript
SingleSelect('#my-select', {}, [
	{ value: 'nl', label: 'Netherlands', selected: true },
	{ value: 'be', label: 'Belgium' }
]);
```

**Notes:**

- To simulate the same behaviour as above, we set `selected: true` for the first object. This property/value is not returned when a user selects an option.

### Adding data after async call

``` markup
<select id="my-select"></select>
```

``` javascript
var singleSelect = SingleSelect('#my-select');
jQuery.get('/endpoint', function(response) {
	singleSelect.setData(response.data);
	singleSelect.render();
	singleSelect.setSelectedRecord(response.data[0]);
});
```

**Notes:**

- The drop down needs to be rendered again after setting data.
- To simulate the same behaviour as the previous two examples, we use `setSelectedRecord` to the first record in the data array. This is not mandatory ofcourse.
- Ofcourse, if the single select was not initialized before the API call, you could simplify this code by passing the `response` as data during initialization in the callback: `SingleSelect('...', {}, response.data)`.

## API

### clearAutocomplete(token)

**Description:** Clears the autocomplete; removes all data, displays 'No results' messsage and opens the menu. This function should normally be used after an API call that produces no results.

**Returns:** _no return value_

| Parameter    | Type     | Description | 
| ------------ | -------- | ----------- | 
| token        | number   | _(optional)_ The token that was passed with the `onAutocompleteQuery` callback. If omitted, will not check for a token and clear the single select anyway; otherwise, will only clear if the instance’s 'current' token matches the passed token. |

``` javascript
singleSelect.element.on('onAutocompleteQuery', function(event, data) {
	// ... async implementation
	...function(response) {
		singleSelect.clearAutocomplete(data.token);
	});
});
```

### disable()

**Description:** Disable a single select. User interaction is no longer possible; click events are canceled. All other functionality, including API interaction, stays intact. Placeholder is removed to signal to the user that there is no input possible.

**Returns:** The single select instance (`this`)

``` javascript
var singleSelect = SingleSelect('[data-single-select]');
singleSelect.disable();
```

### enable()

**Description:** Enable a disabled single select, reallowing user interaction. Also reinstates any configured placeholder.

**Returns:** The single select instance (`this`)

``` javascript
var singleSelect = SingleSelect('[data-single-select]', { disabled: true });
singleSelect.enable();
```

### getSelection()

**Description:** Gets the selected item in the drop down

**Returns:** an object that is an exact clone of the object passed as data to the single select. If you used `<option>`s to set data, the object will be an id/label pair.

**Example:**

``` markup
<select data-single-select>
	<option value="nl" selected>Netherlands</option>
	<option value="be">Belgium</option>
</select>
```

``` javascript
var singleSelect = SingleSelect('[data-single-select]');
singleSelect.getSelection(); // will return { value: 'nl', label: 'Netherlands' }
```

### render()

**Description:** Renders the single select with all 'chrome' (meaning icons, HTML elements etc) and its menu items

**Returns:** _no return value_

``` javascript
var singleSelect = SingleSelect('[data-single-select]').render();
```

### setAutocompleteResult(token, data)

**Description:** The same functionality as `setData`, but should be used after an async call in combination with the `onAutocompleteQuery` event (which provides the token you need to pass back with `setAutocompleteResult`).

**Returns:** _no return value_

| Parameter    | Type     | Description | 
| ------------ | -------- | ----------- | 
| token | number | The token that was passed with the `onAutocompleteQuery` callback. If it does not match the 'current' token, no data will be set |
| data | array | Data as an array, usually holding several object literals representing the drop down choices |

**Example:**

``` javascript
singleSelect.element.on('onAutocompleteQuery', function(event, data) {
	// ... async implementation
	...function(response) {
		singleSelect.setAutocompleteResult(data.token, response);
	});
});
```

### setData(data)

**Description:** Sets data of the single select which makes up the items of the drop down menu. 

**Returns:** _no return value_

| Parameter    | Type     | Description | 
| ------------ | -------- | ----------- | 
| data         | array   | Data as an array, usually holding several object literals representing the drop down choices |

**Example:**

``` javascript
var singleSelect = SingleSelect('[data-single-select]').render();
var data = [{ id: 1, value: 'foo' }, { id: 2, value: 'bar' }];
singleSelect.setData(data);
```

### setSelectedRecord(object)

**Description:** Selects a record in the dropdown. Uses the `keyReference` setting to figure out which item belongs to the object you pass.

**Returns:** a `menuItem` object (representing an item in the dropdown) or _false_ if the passed object could not found.

| Parameter    | Type     | Description | 
| ------------ | -------- | ----------- | 
| item         | object   | The object corresponding to the menu item in the dropdown that should be selected. Usually the exact same object as retrieved from the API. | 

**Example:**

``` javascript
var singleSelect = SingleSelect('[data-single-select]', { keyReference: 'id' });
var data = [{ id: 1, value: 'foo' }, { id: 2, value: 'bar' }];
singleSelect.setData(data);
singleSelect.render();
singleSelect.setSelectedRecord(data[1]); // selects drop down item with id 2, value 'bar'
```

## Events

**Note** that since events can only be called on jQuery objects, and the single select instance itself is not a jQuery object, you should listen to events on the `element` property.

### onAutocompleteQuery(event, data)

**Description:** Fired when an autocomplete query is entered and either the timeout before autosubmitting has passed, or the user pressed [Enter]

**Triggered on:** single select's `element` property

| Parameter    | Type     | Description | 
| ------------ | -------- | ----------- | 
| event | event | The jQuery event | 
| data | object | Data, see below | 

**Data passed:**

| Property    | Type     | Description | 
| ------------ | -------- | ----------- | 
| query | string | The query the user typed in |
| maxResults | number | The maximum number of results the API should return. This is the same as the setting `maxSearchResults` |
| token | number | A unique token that should be used to check if the data returned belongs to the right event | 

**Example:**

``` javascript
var singleSelect = SingleSelect(/* ... */).render();
singleSelect.element.on('onAutocompleteQuery', function(event, data) {
	$.get('/endpoint/?q=' + data.query + '&n=' + data.maxResults, function(response) {
		singleSelect.setAutocompleteResult(data.token, response);
	});
});
```

### onItemChange(event, data)

**Description:** Fired when an item is selected in the dropdown

**Triggered on:** single select's `element` property

| Parameter    | Type     | Description | 
| ------------ | -------- | ----------- | 
| event | event | The jQuery event | 
| data       | object   | The object corresponding to the menu item in the dropdown that should be selected. This should be the same object that was put into the single select. | 

**Example:**

``` javascript
var singleSelect = SingleSelect(/* ... */).render();
singleSelect.element.on('onItemChange', function(event, data) {
	console.log('User picked ' + data.value);
});
```

**Comment by thadeetrompetter** [comment]
The amount of information contained is indeed minimal, but IMO just right for a component that does no need to fulfil the swiss army knife role but just has a couple of config attributes that its consumer needs to know about, as well as a short and concise description. I'd say keep the format as presented in this PR and link to a more verbose example for when things get out of hand.

**Comment by phortuin** [comment]
Agree @thadeetrompetter but can we then at least have the tag attributes in a table, for readability?

**Comment by jbmoelker** [comment]
@phortuin, @thadeetrompetter I agree a table would be much better for the tag attributes. @phortuin you want to make the change? And my hope is that the tag demos would serve as illustrations for the different configurations of a Riot tag, so you wouldn't need to give those examples in the README itself.

**Comment by thadeetrompetter** [comment]
@phortuin: I concur

**Comment by jbmoelker** [comment]
@phortuin updated to use table. Do you approve? I can do merge.

**Comment by phortuin** [comment]
Would prefer not to have fat code (**`on-end`** ) if you don’t mind :grimacing: fat, as in bold

**Comment by jbmoelker** [comment]
@phortuin attributes are now regular weight.

**Comment by phortuin** [comment]
OK i can finally concur :dancer:

**Pull request merged**